packages/application-shell/src/components/login-sso/login-sso.js

```diff
@@ -32,9 +32,17 @@ export const getMessageKeyForError = error => {
   }
 };
 
-function generateAndCacheNonce() {
+function generateAndCacheNonceWithState(state) {
   const nonce = uuid();
-  storage.put(STORAGE_KEYS.NONCE, nonce);
+  // We store additional information within the given `nonce`
+  // to then retrieve it later when the IdP redirects back
+  // to our application. The URL will contain the `nonce` within
+  // the id_token and, once validated, we can retrieve and use
+  // the state object.
+  // https://auth0.com/docs/protocols/oauth2/oauth-state#how-to-use-the-parameter-to-restore-application-state
+  storage.put(`${STORAGE_KEYS.NONCE}_${nonce}`, state, {
```

Comment:
Nice. The nounsr is important here. Is there any way we should clean up old entries? Like once you set a new one here all with a certain pattern are pruned?

Reply:
The entry is cleaned when the login callback is rendered. Also, since I put it now in `sessionStorage`, is automatically cleaned up when the browser is closed.
I think it's enough...